makefiles/cflags.inc.mk

```diff
@@ -57,6 +57,25 @@ ifeq ($(LTO),1)
   LINKFLAGS += $(LTOFLAGS) -ffunction-sections -fdata-sections
 endif
 
+# Cannot test with `BUILD_IN_DOCKER=1`, as this is only the case when the
+# actual build is done in the docker container and we are still running in the
+# host's context.
+ifeq (1,$(BUILD_IN_DOCKER))
+  LINKER_SUPPORTS_NOEXECSTACK := determine-later-inside-docker
+endif
+
+# Check if linker supports `-z noexecstack`. Handle BUILD_IN_DOCKER separately,
+# as this is run in the host environment rather than inside the container. We
+# just hardcode this in the BUILD_IN_DOCKER case for now.
+LINKER_SUPPORTS_NOEXECSTACK ?= $(shell echo "int main(){} void _exit(int n) {(void)n;while(1);}" | LC_ALL=C $(LINK) -xc - -o /dev/null -lc -Wall -Wextra -pedantic -z noexecstack 2> /dev/null && echo 1 || echo 0)
+
+# As we do not use nested functions or other stuff requiring trampoline code,
+# we can safely mark the stack as not executable. This avoids warnings on newer
+# toolchains.
+ifeq (1,$(LINKER_SUPPORTS_NOEXECSTACK))
```

Comment:
Hm so this is effectively always false on CI (because it's `determine-later-inside-docker`, not `1`) - so we'll never build with `-z noexecstack`?

Reply:
No, the build system evaluates twice in the `BUILD_IN_DOCKER` case - which in part explains the long start up time. You can add an

``` Makefile
$(warning LINKER_SUPPORTS_NOEXECSTACK="$(LINKER_SUPPORTS_NOEXECSTACK)")
```

to the file and see it outputing twice, e.g.

```
$ make BUILD_IN_DOCKER=1 -C examples/default/ BOARD=samr21-xpro
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/default'
/home/maribu/Repos/software/RIOT/makefiles/cflags.inc.mk:72: LINKER_SUPPORTS_NOEXECSTACK="determine-later-inside-docker"
Launching build container using image "docker.io/riot/riotbuild:latest".
podman run --rm --tty --userns keep-id -v '/etc/zoneinfo/Europe/Berlin:/etc/localtime:ro' -v '/home/maribu/Repos/software/RIOT:/data/riotbuild/riotbase:delegated' -v '/home/maribu/.cargo/registry:/data/riotbuild/.cargo/registry:delegated' -v '/home/maribu/.cargo/git:/data/riotbuild/.cargo/git:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'    -v '/home/maribu/Repos/software/boards/riot:/data/riotbuild/external/riot:delegated' -v '/home/maribu/Repos/software/miot-pcbs/RIOT/boards:/data/riotbuild/external/boards:delegated'  -e 'BOARD=samr21-xpro' -e 'DISABLE_MODULE=' -e 'DEFAULT_MODULE=' -e 'FEATURES_REQUIRED=' -e 'FEATURES_BLACKLIST=' -e 'FEATURES_OPTIONAL=periph_rtc' -e 'USEMODULE=auto_init_gnrc_netif gnrc gnrc_pktdump gnrc_txtsnd netdev_default ps saul_default shell shell_cmds_default' -e 'USEPKG='  -w '/data/riotbuild/riotbase/examples/default/' 'docker.io/riot/riotbuild:latest' make 'BOARD=samr21-xpro'  'EXTERNAL_BOARD_DIRS=/data/riotbuild/external/riot /data/riotbuild/external/boards'  
/data/riotbuild/riotbase/makefiles/cflags.inc.mk:72: LINKER_SUPPORTS_NOEXECSTACK="1"
[...]
```

or

```
$ make BUILD_IN_DOCKER=1 -C examples/default/ BOARD=z1
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/default'
/home/maribu/Repos/software/RIOT/makefiles/cflags.inc.mk:72: LINKER_SUPPORTS_NOEXECSTACK="determine-later-inside-docker"
Launching build container using image "docker.io/riot/riotbuild:latest".
podman run --rm --tty --userns keep-id -v '/etc/zoneinfo/Europe/Berlin:/etc/localtime:ro' -v '/home/maribu/Repos/software/RIOT:/data/riotbuild/riotbase:delegated' -v '/home/maribu/.cargo/registry:/data/riotbuild/.cargo/registry:delegated' -v '/home/maribu/.cargo/git:/data/riotbuild/.cargo/git:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'    -v '/home/maribu/Repos/software/boards/riot:/data/riotbuild/external/riot:delegated' -v '/home/maribu/Repos/software/miot-pcbs/RIOT/boards:/data/riotbuild/external/boards:delegated'  -e 'BOARD=z1' -e 'DISABLE_MODULE=' -e 'DEFAULT_MODULE=' -e 'FEATURES_REQUIRED=' -e 'FEATURES_BLACKLIST=' -e 'FEATURES_OPTIONAL=periph_rtc' -e 'USEMODULE=auto_init_gnrc_netif gnrc gnrc_pktdump gnrc_txtsnd netdev_default ps saul_default shell shell_cmds_default' -e 'USEPKG='  -w '/data/riotbuild/riotbase/examples/default/' 'docker.io/riot/riotbuild:latest' make 'BOARD=z1'  'EXTERNAL_BOARD_DIRS=/data/riotbuild/external/riot /data/riotbuild/external/boards'  
/data/riotbuild/riotbase/makefiles/cflags.inc.mk:72: LINKER_SUPPORTS_NOEXECSTACK="0"
[...]
```

I think the reason to evaluate twice is so that it can mount only external module dirs, external board dirs, and external package dirs into the container that are actually used. (I'm not entirely sure that this is a good trade-off. Just mounting all of them into the container would likely gain some significant speedup if this allows the build system to be evaluated only once.)